### ISSUE TITLE
Wasm local filename extensions never unloaded when EnvoyFilter deleted

### ISSUE THREAD

**Issue by esnible** [istio/istio#26492]
I suspect this is an Envoy bug.  I am reporting it here because I can recreate the problem in a few lines with an EnvoyFilter.  I don't know Envoy will enough to reproduce w/o Istio.

[wasm-unload.zip](https://github.com/istio/istio/files/5072046/wasm-unload.zip)

To reproduce:
```
unzip wasm-unload.zip
# heartbeat.wasm is my own AssemblyScript plugin compiled with <https://github.com/solo-io/proxy-runtime>
kubectl create configmap new-filter --from-file=new-filter.wasm=heartbeat.wasm
# Deploy a pod in an auto-injected namespace
kubectl apply -f httpbin-sidecar-wasm.yaml
# Create an EnvoyFilter with workload selector `app=httpbin`
kubectl apply -f httpbin-ef.yaml
# Note that if you check :15000/config_dump at this point, "new-filter.wasm" is referenced in six places.
# (That could be my EnvoyFilter, I didn't research it exactly how it attached.)
kubectl logs deployment/httpbin -c istio-proxy -f
# heartbeat.wasm calls calls `proxy_set_tick_period_milliseconds` to be woken up every 10 seconds
# Note that `proxy_on_vm_start` must have been called three times, every 10 seconds it logs three times
# (Why not six times?  I don't know.)
kubectl delete -f httpbin-ef.yaml
```

At this point, `kubectl exec service/httpbin -c istio-proxy -- curl localhost:15000/config_dump` references "new-filter.wasm" ZERO times.  Yet according to `kubectl logs deployment/httpbin -c istio-proxy -f` every 10 seconds it is still logging three times.

I also expected ABI `proxy_on_done` to be called by Envoy and my WASM to log something, but the log never appeared.

**Comment by kyessenov** [comment]
cc @jplevyak @PiotrSikora @mandarjog yet another lifecycle issue.

**Comment by kyessenov** [comment]
Are you getting onDestroy method callback? The VM lives independently from xDS resources. Perhaps, because the timer is not removed, the VM is still ticking.

**Comment by esnible** [comment]
> Are you getting onDestroy method callback? The VM lives independently from xDS resources. Perhaps, because the timer is not removed, the VM is still ticking.

@kyessenov I am using https://github.com/solo-io/proxy-runtime/blob/master/assembly/exports.ts which has no OnDestroy, just `onDone()` and `onDelete()`.  Neither seems to be firing for me.

I'm using Istio 1.7.0 RC1 with Envoy version `73a1449f3bdecea448c28eabac126f23d630abf5/1.15.0/Clean/RELEASE/BoringSSL` (whatever that means!)

Is there a place to look for the Envoy WASM ABI calls that is different from language runtime?  Somewhere in the Envoy tree?

Is there any way for my WASM code to find out what the Envoy version or V8 VM version is being used to run it?

**Comment by mandarjog** [comment]
@yuval-k is typescript support updated  and tested with istio 1.7 or is Ed doing it for the first time?

**Comment by yuval-k** [comment]
i know the spec was changed in 1.7; I'm planning to update it as my next task

**Comment by yuval-k** [comment]
@esnible if you'll provide your source code and can try and test it myself with the updated sdk

**Comment by esnible** [comment]
@yuval-k Here is my source code:

[add-header-and-tick.ts.txt](https://github.com/istio/istio/files/5105334/add-header-and-tick.ts.txt)

This code doesn't do anything useful.  I wrote it to get an understanding of the lifecycle events that a filter might experience.  There may be some errors -- for example, I defined an `onDone()` and `onComplete()` for both my `Context` and my `RootContext`, because I was uncertain if the system would call both, and in what order.  (Currently it is calling neither).

**Comment by yuval-k** [comment]
I'm updating the runtime to sdk version 0.2.0 here:
https://github.com/solo-io/proxy-runtime/pull/22

note that you configuration has an error:
the filter name should be `envoy.filters.http.wasm` instead of `myheartbeat`

**Comment by kyessenov** [comment]
@yuval-k The name in the filter config does not have to be a well known name. The typed config decides which filter is used.

**Comment by yuval-k** [comment]
@esnible I got your code working with the latest code; see attached.
I will publish a new release for the asm-script runtime later today.

https://gist.github.com/yuval-k/5953656537e10a0de214f482719301e0

**Comment by yuval-k** [comment]
We just published 0.1.8 to npm; let me know if you need any other help from me on this.

**Comment by esnible** [comment]
@yuval-k I don't understand your answer.  I will try your gist.

Even with incorrect code, it should be impossible to use set_tick_period_milliseconds() to keep WASM running inside Envoy after the XDS has said that WASM is no longer part of the Envoy configuration.  Envoy should always unload .wasm that is not referenced by any current XDS envoy.extensions.filters.http.wasm.v3.Wasm, even if the original .wasm was purposely malicious.

**Comment by yuval-k** [comment]
i'm not sure about the original bug - i just made sure that the asm-script sdk is updated to support the 0.2.0 ABI spec that's in istio 1.7

**Comment by esnible** [comment]
@yuval-k Thank you for updating https://github.com/solo-io/proxy-runtime
I am struggling to find the documentation.  I see your PR https://github.com/solo-io/proxy-runtime/commit/b545be086fa58ea2b1634afe0423fbfaaa118195#diff-fbe5c7809cca9cce116e35fc93176debR13 updated the ABI version from `proxy_abi_version_0_1_0()` to `proxy_abi_version_0_2_0()`.

Can you explain where to find the docs for these ABI versions?  I expected the docs to be in https://github.com/proxy-wasm/spec/tree/master/abi-versions but that only contains _vNEXT_, not 0.10 or 0.2.0.

The ABI you checked in has a method `proxy_send_local_response()` but the ABI in vNEXT has `proxy_send_http_response()` instead.  Where do I find the doc giving the parameters for proxy_send_local_response?

**Comment by yuval-k** [comment]
@esnible I don't have a good answer here - the route i went is:
- look at istio/proxy@1.7 and  see the [repo/version of envoy it depends on](https://github.com/istio/proxy/blob/1.7.0/WORKSPACE#L55)
- go there and see what version of the wasm host/sdk [it depends on](https://github.com/istio/envoy/blob/c82318df901d0f39018802cf2c953f4b7bed1bb9/bazel/repository_locations.bzl#L490)
- go to the sdk source code and figure out the ABI

**Comment by esnible** [comment]
This should not be closed unless we either verify that

- it is possible to remove WASM extensions added through Istio by deleting them, or 
- this bug is replaced by a bug in the Envoy repo.

@PiotrSikora ?

**Comment by esnible** [comment]
Closing.  Verified that the onTick() stops being serviced after 1 minute.  This is strange ... I expected onTick() to stop receiving service instantly, but 1 minute is reasonable for shutting down the plugin.
